### PR TITLE
Fix collapsing reduction crashes when an input has a broadcast #12211

### PR DESCRIPTION
**Pull request by grypp**
This PR fixes the problem defined in #12211.

In this cas collapsing different dimensional reduction inputs requires inserting control-flow in the loop, therefore we don't collapse that.

### CODE REVIEW

**Comment by MaheshRavishankar** [comment]
Thanks @grypp ! I forgot we have two different collapse passes. Can we fold this reduction only pass into the other collapsing pass.

**Comment by grypp** [comment]
I think these two passes are run in different order. I can take a look whether I can merge them run at the same time.

In the meantime, should I commit the fix to unblock the bug?

**Comment by ThomasRaoux** [comment]
I believe collapsing may cause some ordering issues as for some of the work we did we wanted reduction collapsing to happen before splitReduction pass. I would prefer we don't merge those at the time.

**Review by ThomasRaoux — APPROVED** [review]

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/test/collapse_reduction.mlir
Comment (okkwon) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/test/collapse_reduction.mlir]: nit: need a new line at the end of the file.

**Comment by grypp** [comment]
I am closing this PR since @okkwon already submitted a similar fix.

**Pull request closed without merging**
